### PR TITLE
custom-sysroot: replace preferLocalBuild with runCommandLocal

### PR DESCRIPTION
**Pull request by Stunkymonkey**
###### Motivation for this change
#28910
i have not tested any compilation. First of I want your thoughts about this approach.

###### Things done

- [ ] Tested using sandboxing ([nix.useSandbox](https://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](https://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [ ] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [21.11 Release Notes (or backporting 21.05 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2111-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

**Comment by andir** [comment]
Please describe the change properly on the commit message.

**Comment by Stunkymonkey** [comment]
@andir could you evaluate again please

**Comment by andir** [comment]
The commit still doesn't state any motivation. I see that the issue is about listing phases. Please explain why this change is related to this issue. 

A good commit message tells the reviewer everything they need to know about the change.

Why is this better? What problem does it solve? Are there alternatives?

I really like the way the kernel folks describe it: https://www.kernel.org/doc/html/latest/process/submitting-patches.html#describe-your-changes

I don't mean to devalue your change but to encourage you to write a concise commit message once that avoid a bunch of research and guessing from others. Think about a case where you come across that change (and it caused issues) wouldn't it be great to see why that was done instead of having to read plenty of other resources?

File: pkgs/build-support/rust/sysroot/default.nix [hunk 1]
```diff
@@ -1,25 +1,21 @@
-{ stdenv, rust, rustPlatform, buildPackages }:
+{ stdenv, rust, rustPlatform, buildPackages, runCommandLocal }:
 
 { shortTarget, originalCargoToml, target, RUSTFLAGS }:
 
 let
-  cargoSrc = stdenv.mkDerivation {
-    name = "cargo-src";
-    preferLocalBuild = true;
-    phases = [ "installPhase" ];
```
Comment (SuperSandro2000) [inline pkgs/build-support/rust/sysroot/default.nix:9]: Basically removes the usage of phases and uses a more fitting build command.

**Comment by Stunkymonkey** [comment]
the core motivation was to get rid of `phases`.
`phases` itself is a whitelisting procedure when building packages.
this has one major issue: forgetting `fixupPhase`

there is a new approach by blacklisting every single phase, that is not needed. e.g. `dontConfigure = true;`

i could use the blacklisting here, but this would produce a lot of lines to prevent the usage of `phases. Therefore I did some thinking and cam up with the `runCommandLocal`, which does nearly the same thing like `preferLocalBuild` and additionally executes the few lines of code.

So overall the process does not really change, except there are fewer lines and `phases` is removed.

**Comment by mohe2015** [comment]
@Stunkymonkey Now put this into the commit message after the first line and I think you're good. (Probably not needed that exactly - saying you wanted to get rid of `phases` is probably enough)

**Comment by Stunkymonkey** [comment]
anything i can help with?

**Comment by Stunkymonkey** [comment]
to have an open PR without recent responses does not feel best. Any idea on how to proceed?

**Comment by Stunkymonkey** [comment]
@figsoda maybe you could test this PR same as you did with #134899

**Comment by Stunkymonkey** [comment]
any chance this can get merged eventually?

**Pull request closed without merging**